### PR TITLE
Add support for sampling multi-bucket aggregations, scaling values, and significant terms/text

### PR DESCRIPTION
**Pull request by benwtrent**
This commit completes initial value scaling and sampling support.

The following aggregations are added to sampling support and their doc_counts, other related count values, and sub aggregations scaled appropriately:

- Adjacency matrix  
-  Auto-interval date histogram
- Categorize text  
- Composite
- Date histogram  
- Date range
- Filter
- Filters
- Geo-distance
- Histogram
- IP Range
- Multi-terms
- Range
- Rare terms
- Significant terms*
- Significant text*
- Terms
- Variable width histogram 

* For significance lookup, additional code was added to ensure that the background set applies the same random sub-sampling that the foreground set uses per shard. This is so statistically, both see the same random subset of docs for more intuitive, sound, and explainable results.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java [hunk 1]
```diff
@@ -290,6 +290,10 @@ protected Aggregator doCreateInternal(Aggregator parent, CardinalityUpperBound c
             // counting
             bucketCountThresholds.setShardSize(BucketUtils.suggestShardSideQueueSize(bucketCountThresholds.getRequiredSize()));
         }
+        // We should scale the shard min doc count to account for the reduction due to sampling
+        getSamplingContext().ifPresent(
+            samplingContext -> bucketCountThresholds.setShardMinDocCount(samplingContext.scale(bucketCountThresholds.getShardMinDocCount()))
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorFactory.java:295]: I'd like to have some tests around this.  Terms does non obvious things with `ShardMinDocCount` (see https://github.com/elastic/elasticsearch/issues/70664 for example) and I want to make sure this honors that behavior.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/SamplingContext.java [hunk 2]
```diff
@@ -73,4 +82,22 @@ public double inverseScale(double value) {
         return value;
     }
 
+    public Query buildQueryWithSampler(QueryBuilder builder, AggregationContext context) throws IOException {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/SamplingContext.java:85]: Javadoc on both of these please.  They have similar names but do different things.

**Review by not-napoleon — APPROVED** [review]
Looks good overall, few nits.  As mentioned in slack, I think `scale` and `inverseScale` are potentially confusing, and would suggest `scaleDown` and `scaleUp`, to make it clearer which does what.  No need to block merging on that though, we can do it in another PR, or I'm open to feedback if you don't like the naming.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificanceLookup.java [hunk 3]
```diff
@@ -60,19 +61,27 @@
...
+            ? samplingContext.buildSamplingQueryIfNecessary(context).orElse(null)
+            : samplingContext.buildQueryWithSampler(backgroundFilter, context);
         /*
          * We need to use a superset size that includes deleted docs or we
          * could end up blowing up with bad statistics that cause us to blow
          * up later on.
          */
         IndexSearcher searcher = context.searcher();
-        supersetNumDocs = backgroundFilter == null ? searcher.getIndexReader().maxDoc() : searcher.count(this.backgroundFilter);
+        // TODO can we cheat the random sampler and take a fraction of the max doc?
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificanceLookup.java:83]: I thought more about the statistics here, and I don't think we can reliably take a fraction of the max doc here :/. I will revisit this TODOs as most of them were written back when we were experimenting about how to implement this.

**Pull request merged**